src/decisionengine_modules/NERSC/sources/NerscSFApi.py

```diff
@@ -0,0 +1,146 @@
+import os
+
+import jwt
+import pandas as pd
+import pem
+import requests
+
+# from decisionengine_modules.NERSC.util import newt
+# from datetime import datetime
+from authlib.integrations.requests_client import OAuth2Session
+from authlib.oauth2.rfc7523 import PrivateKeyJWT
+
+from decisionengine.framework.modules import Source
+from decisionengine.framework.modules.Source import Parameter
+
+
+@Source.supports_config(Parameter("constraints", type=dict, comment="""Supports the layout:"""))
+@Source.produces(Nersc_Allocation_SFAPI=pd.DataFrame)
+class NerscSFApi(Source.Source):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.constraints = config.get("constraints")
+        if not isinstance(self.constraints, dict):
+            raise RuntimeError("constraints should be a dict")
+
+        self.logger = self.logger.bind(
+            class_module=__name__.split(".")[-1],
+        )
+        self.localmap = {"uscms": "m2612", "fife": "m3249"}
+        self.keys_list = ["hours_given", "hours_used", "id", "project_hours_given", "project_hours_used", "repo_name"]
+
+    def check_accesstoken(self, nersc_user):
+        token_url = "https://oidc.nersc.gov/c2id/token"
+        #        currenttime = time.time()
+        renew_bool = False
+
+        rawfile_ucms = "/tmp/ucms_access.token"
+        rawfile_fife = "/tmp/fife_access.token"
+
+        # These 4 variables are defined in /etc/decisionengine/config.d/Nersc.jsonnet
+        #          "ucms_private_key": "/etc/decisionengine/modules.data/ucms_privkey.pem",
+        #          "fife_private_key": "/etc/decisionengine/modules.data/fife_privkey.pem",
+        #          "ucms_client_id":   "/etc/decisionengine/modules.data/ucms_clientid.txt",
+        #          "fife_client_id":   "/etc/decisionengine/modules.data/fife_clientid.txt"
+        # And the actual key files and id files reside in /etc/decisionengine/modules.data/
+        # So, puppet should copy these 4 files from
+        # ssiadmin4:/srv/secrets/per_host/hepcsvc03.fnal.gov/credentials/
+        pemfile_ucms = self.constraints.get("ucms_private_key")
+        pemfile_fife = self.constraints.get("fife_private_key")
+        client_id_ucms = self.constraints.get("ucms_client_id")
+        client_id_fife = self.constraints.get("fife_client_id")
+
+        rawfile = None
+        pemfile = None
+        client_id = None
+        if nersc_user == "uscms":
+            rawfile = rawfile_ucms
+            pemfile = pemfile_ucms
+            with open(client_id_ucms) as ifile:
+                client_id = ifile.read()
+                client_id = client_id.rstrip()
+        elif nersc_user == "fife":
+            rawfile = rawfile_fife
+            pemfile = pemfile_fife
+            with open(client_id_fife) as ifile:
+                client_id = ifile.read()
+                client_id = client_id.rstrip()
+        else:
+            self.logger.error("Unknown user, exiting")
+            return None
```

Comment:
Instead of hardcoding the user names, could these be a dictionary in the config file? And for each user you have the keys private_key and client_id and if `nersc_user` is not a valid key in the dictionary then you have the unknown user error.
This way there would be no need to change the code when a user is added, only the configuration.
Unless there is a strong reason for the hardcoding that I'm missing I think it is an important change.

Reply:
I did not see any change in the code or any comment to my request not to hardcode the VO names in the function but to make it more general and leave the VO list in the configuration.
If you think the hardcoding is the best solution, please add a comment stating so.

Reply:
Sorry, I missed this comment from you.
I am testing your idea now.
Will update soon.

Reply:
Okay, I updated the code to move these strings to the config file
While at it, I also moved p_url to the config file.